python/pyarrow/array.pxi

```diff
@@ -497,10 +497,15 @@ cdef class Tensor:
         self.type = pyarrow_wrap_data_type(self.tp.type())
 
     def __repr__(self):
+        if self.tp is NULL:
```

Comment:
Having `__repr__` raise isn't really nice, because it breaks debugging. It would be better to return something like `<invalid pyarrow.Tensor>`. Also you probably want to protect other methods, and raise there if the object isn't initialized.

Reply:
This is really a stopgap until we can replace our Cython API with pybind11. Cython's inability to deal with `shared_ptr` is huge burden right now. We have all these handwritten checks to make sure that an object is valid, which would be completely unnecessary if we moved to pybind.

In any event, I'll add these checks here so we can get this merged.

Reply:
Is pybind really an option? It seems more likely we would migrate bindings to plain C extensions so that we can develop a more mature public C API for pyarrow

Reply:
I would think there's much less boilerplate to write a pure C API + pybind, than to write a pure C API + C extensions. pybind [supports numpy](http://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#) as well, hiding a lot of the complexity of the C APIs behind the guarantees provided by C++ RAII, objects, and templates.

The pure C API would look the same regardless, it's really just a question of whether we want to take advantage of the convenience of pybind, or hand roll extensions where we would have to deal with reference counting and numpy's C API.

Reply:
I have experience both with Cython and the Python C API, Cython is a much more reasonable choice to me. The effort spent on comparable features is easily 2x or 3x larger when writing C code against the CPython API (and the opportunity for bugs is also much higher, given you have to deal with refcounting and GC details by hand). Furthemore, Cython makes it easy to use high-level Python features that are a major pain to emulate in plain C.

Just my 2 cents :-)

Reply:
Besides, aren't the hand-written checks mandated by the current constructor signature and the fact that you have to go through a classmethod to create a proper instance of each Cython wrapper class? If the constructor took the C++ `shared_ptr` as argument and checked its validity, you wouldn't need to sprinkle checks in the other methods/properties.

Reply:
> If the constructor took the C++ shared_ptr as argument and checked its validity, you wouldn't need to sprinkle checks in the other methods/properties.

With pybind the situation is even better, because it would allow us to have constructors for numpy arrays and python lists with the same API e.g., `pa.Tensor([1])`/`pa.Tensor(np.array([1]))` without having to deal with initialization by hand at all.

Reply:
Thus completely hiding the fact that there's a `shared_ptr` in play from Python users.

Reply:
To be clear, I'm advocating for the replacement of Cython with pybind11.

Reply:
I'll open a JIRA if there isn't already one, and start a mailing list discussion. GitHub is getting a bit chatty.

Reply:
Currently, pyarrow has a _public_ Cython and C++ API. If pybind does not support creating public C/C++ API for thirdparty libraries to expose its extension types to non-Python code, it is a non-starter

Reply:
e.g. https://github.com/apache/arrow/blob/master/python/pyarrow/tests/pyarrow_cython_example.pyx

Reply:
Looks like it's pretty straightforward to go back and forth over that boundary https://github.com/pybind/pybind11/blob/master/docs/advanced/pycpp/object.rst#casting-back-and-forth

Reply:
Let's take up some prototyping in a separate patch or repo to understand what a pybind11-based C++ API for pyarrow would look like or how it would work. This is already being used in turbodbc (which uses pybind11 for its bindings -- see https://github.com/blue-yonder/turbodbc/blob/master/cpp/turbodbc_arrow/Library/src/arrow_result_set.cpp#L252)

Reply:
For sure. Didn't mean to derail this conversation.